### ISSUE TITLE
how to access backbuffer ram/texture of renderer directly ?

### ISSUE THREAD

**Issue by rofl0r** [libsdl-org/SDL#9260]
on PSP, doing framebuffer-style drawing is extremely slow.

the usual way is
1) manipulate pixels on some surface or pointer retrieved by SDL_LockTexture
2) SDL_UnlockTexture / SDL_UpdateTexture: entire memory gets copied to texture
3) SDL_RenderCopy/Present texture finally gets copied to vram.

doing 2 additional memcpy()s for a core i7 desktop pc is no big deal, memory is clocked in gigahertz range.

but on PSP, it would be necessary to access VRAM behind renderer/window directly to achieve acceptable performance.

i can't even seem to find a way to directly access the texture's pixels, much less the ones of the window.

and then there's the mismatch between SDL_TEXTUREACCESS_STREAMING and SDL_TEXTUREACCESS_TARGET - you basically want both for PSP, i.e. oring the flags together. and indeed, the code in psp renderer checks using & whereas SDL_LockTexture refuses to work with anything but a plain STREAMING texture.

<strike>a function to read pixels from renderer is provided: SDL_RenderReadPixels, but not the other way.</strike> no direct access either, yet another memcpy.

i'm not talking about complicated setup here, basically a single ram buffer, or texture, representing the entire screen.

**Comment by slouken** [comment]
SDL has an API for this, it's `SDL_GetWindowSurface()`

You basically want to implement the window framebuffer methods the way it's done in src/video/dummy/SDL_nullframebuffer.c, but instead of creating a surface, you'd want to use SDL_CreateSurfaceFrom() and give it the address of the video framebuffer.

**Comment by rofl0r** [comment]
that's very useful info, thanks!

so i'm about to implement
```
int PSP_CreateWindowFramebuffer(_THIS, SDL_Window *window, Uint32 *format, void **pixels, int *pitch)
int PSP_UpdateWindowFramebuffer(_THIS, SDL_Window *window, const SDL_Rect *rects, int numrects)
void PSP_DestroyWindowFramebuffer(_THIS, SDL_Window *window)
```
PSP backend internally uses 2 buffers in vram: frontbuffer and backbuffer, for doublebuffering.
i now could make it so PSP_CreateWindowFramebuffer sets pixels to frontbuffer, and the consumer of SDL_GetWindowSurface() could write directly onto it. but how i could make it so the surface returned alternates between front or backbuffer ?

**Comment by slouken** [comment]
You could change the pixels pointer to point to the correct data each frame. Typically the surface wouldn't actually be VRAM, it was instead be copied to VRAM in PSP_UpdateWindowFramebuffer(), but I don't know what your constraints are in this case.

**Comment by rofl0r** [comment]
apart from my PR, which allows one to paint directly to the FB and achieve 60 fps easily if there are not too many other computations per frame - sufficient for my mp3 player app which just renders a directory listing-, i found that simply turning the check in SDL_LockTexture into a flag test, and then creating a 16 bit window, renderer, and texture with both streaming and target bits set results in the PSP render code putting that texture into VRAM, setting the swizzled flag, for it, while i can access the underlying VRAM directly in my code.

my app (gnuboy) renders to a 16 bit 160x144 buffer each frame and ... *drumroll*... "copying" that buffer onto the texture's VRAM via swizzle_fast and then calling rendercopy/renderpresent results in the GE stretching the texture for me onto the screen in **2 ms**. that includes the time required for the swizzle_fast code. that means you could theoretically achieve something like 4-500 fps if you just keep throwing almost the same pixels at it.

now if you can't combine the 2 access flags like this, you basically need an additional surface to render to (unswizzled) and update the target texture from it, so you fave at least one or 2 additional copies involved, and the backend has to do the swizzling later anyway (if you can even get it to use swizzling there, as the internal ShouldSwizzle function seems a bit suspect).

**Comment by slouken** [comment]
I'm moving this to the ABI milestone, so we can investigate what limitations on the various texture types are. It may be that we make the texture access into flags, or streaming access is always available in some way.

**Comment by slouken** [comment]
So I did some investigating, and D3D11 doesn't allow mixing streaming and render target access. The other renderers work, but implement this by creating staging buffers in CPU memory and returning those in the lock call, so you're not getting direct access to the VRAM that can be addressed by the CPU.

Also, someone using the API naively would naturally expect to be able to lock the surface and get back what has been rendered to the texture, which isn't possible on any other platform.

Hypothetically though, even if we did implement this, it would still be a separate copy since you'd be creating a render texture, locking that, and then copying it to the screen, is that right? Maybe instead it makes sense to have all streaming textures on the PSP placed in VRAM, that way they all have accelerated access? You don't actually need to render to the texture, correct?
